File: packages/core/src/session.ts
import { ACTIVE_TO_REST_FADE_SEC } from "./tone-set.js";
import type { Phase, Session, SessionConfig, SessionEvent } from "./types.js";

export class InvalidSessionConfigError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "InvalidSessionConfigError";
  }
}

function validate(config: SessionConfig): void {
  const positives: Array<keyof SessionConfig> = [
    "inhaleSec",
    "exhaleSec",
    "activeSec",
    "restSec",
    "rounds",
  ];
  for (const key of positives) {
    const value = config[key];
    if (!Number.isFinite(value) || value <= 0) {
      throw new InvalidSessionConfigError(
        `${key} must be a finite positive number, got ${String(value)}`,
      );
    }
  }
  if (!Number.isInteger(config.rounds)) {
    throw new InvalidSessionConfigError(
      `rounds must be an integer, got ${String(config.rounds)}`,
    );
  }
  const breathCycleSec = config.inhaleSec + config.exhaleSec;
  if (breathCycleSec > config.activeSec) {
    throw new InvalidSessionConfigError(
      `activeSec (${config.activeSec}) must fit at least one full breath cycle (inhale + exhale = ${breathCycleSec})`,
    );
  }
}

function computeTotalDurationSec(config: SessionConfig): number {
  return config.rounds * (config.activeSec + config.restSec);
}

function buildSchedule(
  config: SessionConfig,
  startMs: number,
  totalDurationSec: number,
): Array<{ atMs: number; event: SessionEvent }> {
  const scheduled: Array<{ atMs: number; event: SessionEvent }> = [];
  const cycleSec = config.inhaleSec + config.exhaleSec;

  for (let r = 1; r <= config.rounds; r++) {
    const roundStartMs =
      startMs + (r - 1) * (config.activeSec + config.restSec) * 1000;

    // Emit one count event per second within the active phase. Each
    // breath cycle = inhaleSec inhale-counts followed by exhaleSec
    // exhale-counts. We bail out the moment a count would land at or past
    // the active phase boundary so the rest-start event isn't preceded
    // by a stray count.
    for (let c = 0; ; c++) {
      const inhaleOffsetSec = c * cycleSec;
      const exhaleOffsetSec = inhaleOffsetSec + config.inhaleSec;
      if (inhaleOffsetSec >= config.activeSec) break;

      for (let i = 0; i < config.inhaleSec; i++) {
        const offsetSec = inhaleOffsetSec + i;
        if (offsetSec >= config.activeSec) break;
        const atMs = roundStartMs + offsetSec * 1000;
        scheduled.push({
          atMs,
          event: {
            kind: "inhale-count",
            round: r,
            beatIndex: i,
            beatsInPhase: config.inhaleSec,
            atMs,
          },
        });
      }

      if (exhaleOffsetSec < config.activeSec) {
        for (let i = 0; i < config.exhaleSec; i++) {
          const offsetSec = exhaleOffsetSec + i;
          if (offsetSec >= config.activeSec) break;
          const atMs = roundStartMs + offsetSec * 1000;
          scheduled.push({
            atMs,
            event: {
              kind: "exhale-count",
              round: r,
              beatIndex: i,
              beatsInPhase: config.exhaleSec,
              atMs,
            },
          });
        }
      }
    }

    const restAtMs = roundStartMs + config.activeSec * 1000;
    scheduled.push({
      atMs: restAtMs,
      event: {
        kind: "rest-start",
        round: r,
        durationSec: config.restSec,
        fadeOutSec: ACTIVE_TO_REST_FADE_SEC,
        atMs: restAtMs,
      },
    });

    const roundCompleteAtMs = restAtMs + config.restSec * 1000;
    scheduled.push({
      atMs: roundCompleteAtMs,
      event: { kind: "round-complete", round: r, atMs: roundCompleteAtMs },
    });
  }

  const sessionCompleteAtMs = startMs + totalDurationSec * 1000;
  scheduled.push({
    atMs: sessionCompleteAtMs,
    event: { kind: "session-complete", atMs: sessionCompleteAtMs },
  });

  // Stable sort — equal-atMs items preserve insertion order (round-complete before session-complete)
  scheduled.sort((a, b) => a.atMs - b.atMs);

  return scheduled;
}

function phaseFromEvent(event: SessionEvent): Phase | null {
  switch (event.kind) {
    case "inhale-count":
      return event.beatIndex === 0
        ? { kind: "active-inhale", round: event.round, startedAtMs: event.atMs }
        : null;
    case "exhale-count":
      return event.beatIndex === 0
        ? { kind: "active-exhale", round: event.round, startedAtMs: event.atMs }
        : null;
    case "rest-start":
      return { kind: "rest", round: event.round, startedAtMs: event.atMs };
    case "session-complete":
      return { kind: "complete" };
    case "round-complete":
      return null;
    default:
      return null;
  }
}

export function createSession(config: SessionConfig): Session {
  validate(config);
  const totalDurationSec = computeTotalDurationSec(config);

  let phase: Phase = { kind: "idle" };
  let stopped = false;
  let queue: Array<{ atMs: number; event: SessionEvent }> = [];
  // Two cursors over the same queue:
  //   firedCursor — advanced by tick(), drives UI / phase transitions at the
  //                 moment events become audible.
  //   audioCursor — advanced by tickAudio(), drives pre-scheduled audio
  //                 against the engine's clock. Always >= firedCursor.
  // The cursors are re-aligned by rewindAudioCursor() when the host needs to
  // cancel queued audio (e.g. on pause).
  let firedCursor = 0;
  let audioCursor = 0;
  // Pause arithmetic. While paused, effective time freezes at `pausedAtMs`
  // (the nowMs passed to pause()). On resume, totalPausedMs accumulates the
  // wall-clock duration of the pause, so subsequent ticks compute effective
  // time as nowMs - totalPausedMs.
  let pausedAtMs: number | null = null;
  let totalPausedMs = 0;

  function effectiveMs(nowMs: number): number {
    if (pausedAtMs !== null) return pausedAtMs - totalPausedMs;
    return nowMs - totalPausedMs;
  }

  function drain(nowMs: number): readonly SessionEvent[] {
    const effective = effectiveMs(nowMs);
    const events: SessionEvent[] = [];
    while (firedCursor < queue.length) {
      const entry = queue[firedCursor];
      if (!entry || entry.atMs > effective) break;
      firedCursor++;
      events.push(entry.event);
      const next = phaseFromEvent(entry.event);
      if (next !== null) phase = next;
    }
    // The audio cursor must never lag the fired cursor — if a tick consumes
    // events the audio cursor hasn't seen yet (no host using tickAudio, or
    // pre-rewind state), keep them in step.
    if (audioCursor < firedCursor) audioCursor = firedCursor;
    return events;
  }

  function drainAudio(
    nowMs: number,
    lookaheadMs: number,
  ): readonly SessionEvent[] {
    // While paused, effective time is frozen — surface no new events for
    // pre-scheduling. The host has typically just cancelled queued audio.
    if (pausedAtMs !== null) return [];
    const horizon = effectiveMs(nowMs) + lookaheadMs;
    const events: SessionEvent[] = [];
    while (audioCursor < queue.length) {
      const entry = queue[audioCursor];
      if (!entry || entry.atMs > horizon) break;
      audioCursor++;
      events.push(entry.event);
    }
    return events;
  }

  return {
    start(nowMs: number): readonly SessionEvent[] {
      if (stopped) return [];
      queue = buildSchedule(config, 0, totalDurationSec);
      firedCursor = 0;
      audioCursor = 0;
      pausedAtMs = null;
      totalPausedMs = nowMs; // so effective time starts at 0 regardless of nowMs
      return drain(nowMs);
    },

    tick(nowMs: number): readonly SessionEvent[] {
      if (stopped || queue.length === 0) return [];
      return drain(nowMs);
    },

    tickAudio(nowMs: number, lookaheadMs: number): readonly SessionEvent[] {
      if (stopped || queue.length === 0) return [];
      return drainAudio(nowMs, lookaheadMs);
    },

    rewindAudioCursor(): void {
      audioCursor = firedCursor;
    },

    pause(nowMs: number): void {
      if (stopped || pausedAtMs !== null) return;
      pausedAtMs = nowMs;
    },

    resume(nowMs: number): void {
      if (stopped || pausedAtMs === null) return;
      totalPausedMs += nowMs - pausedAtMs;
      pausedAtMs = null;
    },

    stop(): void {
      stopped = true;
      phase = { kind: "idle" };
      queue = [];
      firedCursor = 0;
      audioCursor = 0;
      pausedAtMs = null;
    },

    effectiveMs(nowMs: number): number {
      return effectiveMs(nowMs);
    },

    get phase() {
      return phase;
    },

    get isPaused() {
      return pausedAtMs !== null;
    },

    get totalDurationSec() {
      return totalDurationSec;
    },
  };
}

File: apps/macos/Sources/Breathe/Session.swift
import Foundation

/// Swift port of packages/core/src/session.ts. Same schedule, same events —
/// if we ever spot drift between platforms, this file is the suspect.

struct SessionConfig: Codable, Equatable {
    var inhaleSec: Double
    var exhaleSec: Double
    var activeSec: Double
    var restSec: Double
    var rounds: Int

    static let `default` = SessionConfig(
        inhaleSec: 4, exhaleSec: 4, activeSec: 120, restSec: 15, rounds: 4
    )

    var totalDurationSec: Double {
        Double(rounds) * (activeSec + restSec)
    }

    func formattedDuration() -> String {
        let total = Int(totalDurationSec.rounded())
        let min = total / 60
        let sec = total % 60
        return sec == 0 ? "\(min) min" : "\(min) min \(sec) sec"
    }
}

enum SessionEvent: Equatable {
    case inhaleStart(round: Int, durationSec: Double, atMs: Double)
    case exhaleStart(round: Int, durationSec: Double, atMs: Double)
    case restStart(round: Int, durationSec: Double, fadeOutSec: Double, atMs: Double)
    case roundComplete(round: Int, atMs: Double)
    case sessionComplete(atMs: Double)

    var atMs: Double {
        switch self {
        case .inhaleStart(_, _, let t), .exhaleStart(_, _, let t),
             .restStart(_, _, _, let t), .roundComplete(_, let t),
             .sessionComplete(let t):
            return t
        }
    }
}

struct InvalidSessionConfig: Error { let message: String }

/// The fade duration used at active → rest boundary. Mirrors
/// packages/core/src/tone-set.ts `ACTIVE_TO_REST_FADE_SEC`.
let activeToRestFadeSec: Double = 1.5

final class BreathSession {
    private let config: SessionConfig
    private var queue: [(atMs: Double, event: SessionEvent)] = []
    private var cursor = 0
    private var stopped = false

    init(config: SessionConfig) throws {
        let positives: [(String, Double)] = [
            ("inhaleSec", config.inhaleSec),
            ("exhaleSec", config.exhaleSec),
            ("activeSec", config.activeSec),
            ("restSec", config.restSec),
            ("rounds", Double(config.rounds)),
        ]
        for (name, v) in positives where !v.isFinite || v <= 0 {
            throw InvalidSessionConfig(
                message: "\(name) must be a finite positive number, got \(v)"
            )
        }
        let cycle = config.inhaleSec + config.exhaleSec
        if cycle > config.activeSec {
            throw InvalidSessionConfig(
                message: "activeSec (\(config.activeSec)) must fit one breath cycle (\(cycle))"
            )
        }
        self.config = config
    }

    func start(nowMs: Double) -> [SessionEvent] {
        guard !stopped else { return [] }
        queue = Self.buildSchedule(config: config, startMs: nowMs)
        cursor = 0
        return drain(nowMs: nowMs)
    }

    func tick(nowMs: Double) -> [SessionEvent] {
        guard !stopped, !queue.isEmpty else { return [] }
        return drain(nowMs: nowMs)
    }

    func stop() {
        stopped = true
        queue = []
    }

    private func drain(nowMs: Double) -> [SessionEvent] {
        var out: [SessionEvent] = []
        while cursor < queue.count {
            let entry = queue[cursor]
            if entry.atMs > nowMs { break }
            cursor += 1
            out.append(entry.event)
        }
        return out
    }

    private static func buildSchedule(
        config: SessionConfig, startMs: Double
    ) -> [(atMs: Double, event: SessionEvent)] {
        var out: [(Double, SessionEvent)] = []
        let cycle = config.inhaleSec + config.exhaleSec

        for r in 1...config.rounds {
            let roundStart = startMs + Double(r - 1) * (config.activeSec + config.restSec) * 1000

            var c = 0
            while true {
                let inhaleOffset = Double(c) * cycle
                let exhaleOffset = inhaleOffset + config.inhaleSec
                if inhaleOffset >= config.activeSec { break }

                let inhaleAt = roundStart + inhaleOffset * 1000
                out.append((inhaleAt, .inhaleStart(
                    round: r, durationSec: config.inhaleSec, atMs: inhaleAt
                )))

                if exhaleOffset < config.activeSec {
                    let exhaleAt = roundStart + exhaleOffset * 1000
                    out.append((exhaleAt, .exhaleStart(
                        round: r, durationSec: config.exhaleSec, atMs: exhaleAt
                    )))
                }
                c += 1
            }

            let restAt = roundStart + config.activeSec * 1000
            out.append((restAt, .restStart(
                round: r, durationSec: config.restSec,
                fadeOutSec: activeToRestFadeSec, atMs: restAt
            )))

            let roundCompleteAt = restAt + config.restSec * 1000
            out.append((roundCompleteAt, .roundComplete(round: r, atMs: roundCompleteAt)))
        }

        let sessionCompleteAt = startMs + config.totalDurationSec * 1000
        out.append((sessionCompleteAt, .sessionComplete(atMs: sessionCompleteAt)))

        // Stable sort by atMs — equal-atMs items keep insertion order.
        let indexed = out.enumerated().map { ($0, $1.0, $1.1) }
        let sorted = indexed.sorted { a, b in
            a.1 == b.1 ? a.0 < b.0 : a.1 < b.1
        }
        return sorted.map { ($0.1, $0.2) }
    }
}
